### PR TITLE
randomCode: randomAddresses generation

### PR DESCRIPTION
**Pull request by winsvega**
slowly moving towards generating a random state test

### CODE REVIEW

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 1]
```diff
@@ -509,19 +511,48 @@ void RandomCodeOptions::setWeight(eth::Instruction _opCode, int _weight)
...
-void RandomCodeOptions::addAddress(Address const& _address)
+void RandomCodeOptions::addAddress(Address const& _address, AddressType _type)
 {
-	addressList.push_back(_address);
+	switch(_type)
+	{
+		case AddressType::ACCOUNT:
+			accountAddressList.push_back(_address);
+			callAddressList.push_back(_address);
+		break;
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: The indentation of `break;` look weird to me.  It is just another statement following the previous one.

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 2]
```diff
@@ -302,26 +305,20 @@ std::string RandomCode::generate(int _maxOpNumber, RandomCodeOptions _options)
...
-	return toCompactHexPrefixed((u256)randUInt64Gen() % _maxVal);
+	return toCompactHexPrefixed(RandomCode::randomUniInt(_minVal, _maxVal), 1);
 }
 
-u256 RandomCode::randomUniInt(u256 _maxVal)
+u256 RandomCode::randomUniInt(u256 const& _minVal, u256 const& _maxVal)
 {
-	if (_maxVal == 0)
-		_maxVal = std::numeric_limits<uint64_t>::max();
+	assert(_minVal >= _maxVal);
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: I think it should be `<=` instead of `>=`.

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 3]
```diff
@@ -515,14 +515,14 @@ void RandomCodeOptions::addAddress(Address const& _address, AddressType _type)
...
-		case AddressType::ACCOUNT:
-			accountAddressList.push_back(_address);
-			callAddressList.push_back(_address);
-		break;
 		case AddressType::CALLONLY:
 			callAddressList.push_back(_address);
 		break;
+		case AddressType::ALL:
+		case AddressType::ACCOUNT:
 		default:
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: I think it's better to list all cases
```
case AddressType::ALL:
case AddressType::ACCOUNT:
    something;
    break;
default;
    BOOST_ERROR("unexpected AddressType");
```

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 4]
```diff
@@ -509,19 +511,47 @@ void RandomCodeOptions::setWeight(eth::Instruction _opCode, int _weight)
...
-		int index = (int)RandomCode::randomUniInt() % addressList.size();
-		return addressList[index];
+		case AddressType::CALLONLY:
+			if (RandomCode::randomPercent() < emptyAddressProbability)
+				return ZeroAddress;
+			return callAddressList[(int)RandomCode::randomUniInt(0, callAddressList.size())];
+		case AddressType::ACCOUNT:
+			return accountAddressList[(int)RandomCode::randomUniInt(0, accountAddressList.size())];
+		case AddressType::ALL:
+		default:
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: Here as well.

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 5]
```diff
@@ -509,19 +511,49 @@ void RandomCodeOptions::setWeight(eth::Instruction _opCode, int _weight)
...
-	addressList.push_back(_address);
+	switch(_type)
+	{
+		case AddressType::CALLONLY:
+			callAddressList.push_back(_address);
+			break;
+		case AddressType::ALL:
+		case AddressType::ACCOUNT:
+			accountAddressList.push_back(_address);
+			callAddressList.push_back(_address);
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: Why both lists are modified here?

Comment (winsvega) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: Cause account could be a call address

Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: I will try to figure out what is a call address and what is an account address.

Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: I don't understand.  What can you do on a call address that you cannot do on an account address?

Comment (winsvega) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: an address like 0x0000000...02  is a precompiled address that should be available for call only. there should be no account in pre state for that address

Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: I see. What about `stateAddress` instead of `accountAddress`?

Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: Maybe `accountStateAddress`.

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 6]
```diff
@@ -509,19 +511,49 @@ void RandomCodeOptions::setWeight(eth::Instruction _opCode, int _weight)
...
-	if (addressList.size() > 0)
+	switch(_type)
 	{
-		int index = (int)RandomCode::randomUniInt() % addressList.size();
-		return addressList[index];
+		case AddressType::CALLONLY:
+			if (RandomCode::randomPercent() < emptyAddressProbability)
+				return ZeroAddress;
+			return callAddressList[(int)RandomCode::randomUniInt(0, callAddressList.size())];
+		case AddressType::ACCOUNT:
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: So this should also be `STATE` or `InState`

Comment (winsvega) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: could you please mark all the proposals before I make a commit?

File: test/tools/fuzzTesting/fuzzHelper.h [hunk 7]
```diff
@@ -46,18 +46,27 @@ typedef boost::random::variate_generator<boost::mt19937&, boostUint64 > boostUIn
 struct RandomCodeOptions
 {
 public:
+	enum AddressType{
+		CALLONLY,
+		ACCOUNT,
+		ALL
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.h]: The coding rules say these should be
```
CallOnly,
Account,
All
```
but as I said I think
```
CallOnly,
InState,
All
```
would be better.

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 6 above)
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: That's impossible.   It's a process of approximations.

File: test/tools/fuzzTesting/fuzzHelper.h (hunk 7 above)
Comment (winsvega) [inline test/tools/fuzzTesting/fuzzHelper.h]: why eth Instruction enum all names are CAPITAL then ?

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 8]
```diff
@@ -509,19 +511,49 @@ void RandomCodeOptions::setWeight(eth::Instruction _opCode, int _weight)
...
 {
-	if (addressList.size() > 0)
+	switch(_type)
 	{
-		int index = (int)RandomCode::randomUniInt() % addressList.size();
-		return addressList[index];
+		case AddressType::CALLONLY:
+			if (RandomCode::randomPercent() < emptyAddressProbability)
+				return ZeroAddress;
+			return callAddressList[(int)RandomCode::randomUniInt(0, callAddressList.size())];
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: Currently, `callAddressList` contains all kinds of addresses, not only the CALLONLY addresses.

`AddressType::CallOnly` is in `callOnlyAddressList`.

`AddressType::InState` is in `stateAddressList`.

`AddressType::All` is the union of these.

File: test/tools/fuzzTesting/fuzzHelper.h (hunk 7 above)
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.h]: That's against the coding rule.  Needs to be fixed.

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 8 above)
Comment (winsvega) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: then it would be that question why callOlnly request  return  both inState and CallOnly

Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: I'd ask that.  callOnly request should return callOnly, not both.  For getting both, you should ask for All.

**Pull request merged**